### PR TITLE
[FIXED JENKINS-22665] Fixes for JENKINS-22665 and JENKINS-19755

### PR DESCRIPTION
**Pull request by patbos**
Remove the serialization of the User object within MyUserIdCause. Kept MyUserIdCause for backwards compatibility but is now just a simple subclass of the core class UserIdCause. 
This also fixes the JENKINS-19755 "Manual step "started by" username changes to anonymous after restart" since it now can load the correct user who triggered the build.
After upgrading it is possible to "Manage Old Data" to remove User object serialization of MyUserIdCause.

### CODE REVIEW

**Comment by jglick** [comment]
OK, if `userId` is set in XML then it should work. I would recommend verifying all this with a `@LocalData` test if you can.

**Comment by patbos** [comment]
Added tests for reading old builds and still get the userId and to verify that UserIdCause it used when triggering new build with the manual trigger.

**Comment by patbos** [comment]
Is this ok to merge now?

**Comment by patbos** [comment]
Merged changes from master

File: src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java [hunk 1]
```diff
@@ -24,15 +24,8 @@
  */
 package au.com.centrumsystems.hudson.plugin.buildpipeline;
 
-import hudson.model.Action;
-import hudson.model.FreeStyleBuild;
-import hudson.model.ItemGroup;
-import hudson.model.TopLevelItem;
+import hudson.model.*;
```
Comment (KostyaSha) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java]: See no reason for replacing with single line.

Comment (patbos) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java]: Fixed

**Comment by KostyaSha** [comment]
@patbos see my answer in maillist and please join IRC

File: src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java [hunk 2]
```diff
@@ -341,6 +344,18 @@ public void testHasPermission() throws IOException {
         assertTrue(testView.hasPermission(Permission.READ));
     }
 
+    @Test
+    @LocalData
+    @Bug(19755)
```
Comment (KostyaSha) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java]: `@Bug` is deprecated, replace with `@Issue` if core version allows it

File: src/test/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest/testMyUserIdCauseConversion/config.xml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      <refreshFrequency>3</refreshFrequency>
+      <showPipelineDefinitionHeader>false</showPipelineDefinitionHeader>
+    </au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView>
+  </views>
+  <primaryView>All</primaryView>
+  <slaveAgentPort>0</slaveAgentPort>
+  <label></label>
+  <nodeProperties/>
+  <globalNodeProperties/>
+</hudson>
```
Comment (KostyaSha) [inline src/test/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest/testMyUserIdCauseConversion/config.xml]: Add new lines maybe

Comment (patbos) [inline src/test/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest/testMyUserIdCauseConversion/config.xml]: Fixed

**Comment by KostyaSha** [comment]
And please reword commit to [FIXED JENKINS-22665] description :)

File: src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java (hunk 2 above)
Comment (patbos) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java]: Core version is to low, does not contain the new annotation.

Comment (KostyaSha) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java]: wow..

Comment (KostyaSha) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineViewTest.java]: Yes, appeared only from test-annotations-1.1 :(

**Comment by KostyaSha** [comment]
I mean git commit :) Then it will automatically close jira issue after merge :P

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java [hunk 4]
```diff
@@ -151,71 +149,13 @@
...
-                return false;
-            }
-            if (!(o instanceof Cause.UserIdCause)) {
-                return false;
-            }
-
-            return hashCode() == o.hashCode();
-        }
+    @Deprecated
+    private static class MyUserIdCause extends Cause.UserIdCause {
```
Comment (KostyaSha) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java]: remove empty line...

Comment (patbos) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java]: Fixed

**Comment by patbos** [comment]
Fixed the commit description :)

**Comment by KostyaSha** [comment]
:+1:

**Comment by daniel-beck** [comment]
Caused [JENKINS-27748](https://issues.jenkins-ci.org/browse/JENKINS-27748) (with some assistance by Xstream).

**Comment by dserodio** [comment]
Forgive me for digging up this PR, but is the "converter who converts all au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause to hudson.model.Cause$UserIdCause in the build.xmls" implemented? I'm still seeing problems after upgrading from 1.4.3 to 1.4.8 and I'm fairly sure it's caused by this.

**Comment by KostyaSha** [comment]
You can't update this crappy deserialized xml as it can't load back at all. Just wrote simple script that will physically remove this section from build.xml.

**Comment by dserodio** [comment]
Thanks @KostyaSha should I rip the entire "MyUserIdCause" tag? Do you by any chance have a script ready? :)

**Comment by KostyaSha** [comment]
@dserodio unfortunately i have no script. I suggested it to one of our customers and they resolved this issue. Afair it should be enough to cut one tag (don't remember, probably triple* something).

**Pull request merged**